src/auth/pluggable-auth-handler.ts

```diff
@@ -0,0 +1,219 @@
+    envMap: Map<string, string>
+  ): Promise<ExecutableResponse> {
+    return new Promise((resolve, reject) => {
+      // Spawn process to run executable using added environment variables.
+      const child = childProcess.spawn(
+        this.commandComponents[0],
+        this.commandComponents.slice(1),
+        {
+          env: {...process.env, ...Object.fromEntries(envMap)},
+        }
+      );
+      let output = '';
+      // Append stdout to output as executable runs.
+      child.stdout.on('data', (data: string) => {
+        output += data;
+      });
+      // Append stderr as executable runs.
+      child.stderr.on('data', (err: string) => {
+        output += err;
+      });
+
+      // Set up a timeout to end the child process and throw an error.
+      const timeout = setTimeout(() => {
+        // Kill child process and remove listeners so 'close' event doesn't get
+        // read after child process is killed.
+        child.removeAllListeners();
+        child.kill();
+        return reject(
+          new Error(
+            'The executable failed to finish within the timeout specified.'
+          )
+        );
+      }, this.timeoutMillis);
+
+      child.on('close', (code: number) => {
+        // Cancel timeout if executable closes before timeout is reached.
+        clearTimeout(timeout);
+        if (code === 0) {
+          // If the executable completed successfully, try to return the parsed response.
+          try {
+            const responseJson = JSON.parse(output) as ExecutableResponseJson;
+            const response = new ExecutableResponse(responseJson);
+            return resolve(response);
+          } catch (error) {
+            if (error instanceof ExecutableResponseError) {
+              return reject(error);
+            }
+            return reject(
+              new ExecutableResponseError(
+                `The executable returned an invalid response: ${output}`
+              )
+            );
+          }
+        } else {
+          return reject(new ExecutableError(output, code.toString()));
+        }
+      });
+    });
+  }
+
+  /**
+   * Checks user provided output file for response from previous run of
+   * executable and return the response if it exists, is formatted correctly, and is not expired.
+   */
+  retrieveCachedResponse(): Promise<ExecutableResponse | undefined> {
+    return new Promise((resolve, reject) => {
+      if (!this.outputFile || this.outputFile.length === 0) {
+        return resolve(undefined);
+      }
+
+      let filePath: fs.PathLike;
+      try {
+        filePath = fs.realpathSync(this.outputFile as string);
+      } catch {
+        // If file path cannot be resolved, return undefined.
+        return resolve(undefined);
+      }
+
+      if (!fs.lstatSync(filePath).isFile()) {
+        // If path does not lead to file, return undefined.
+        return resolve(undefined);
+      }
+
+      const readStream = fs.createReadStream(filePath);
```

Comment:
What are your thoughts on using `childProcess.readFile` instead? It may shrink this logic a bit